functions/prepare-check/index.js

```diff
@@ -0,0 +1,99 @@
+const R = require('ramda')
+
+/**
+ * Write to Table Storage for fast pupil authentication
+ * Reads from a queue
+ * @param context
+ * @param {} prepareCheckMessage
+ */
+module.exports = function (context, prepareCheckMessage) {
+  context.log.verbose('prepareCheck: got message from prepare-check queue', prepareCheckMessage)
+  // TODO: Add a version strategy: version field, version handler.
+  // TODO: Add batch processing: e.g. handle 100 at a time
+  try {
+    validateMessage(prepareCheckMessage)
+  } catch (error) {
+    // After 5 attempts at processing the message will be moved to the poison queue
+    // https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-queue#trigger---poison-messages
+    context.log.error('prepareCheck: message failed validation')
+    context.done(error)
+    return
+  }
+
+  context.bindings.preparedCheckTable = []
+
+  const preparedCheck = {
+    partitionKey: prepareCheckMessage.schoolPin,
+    rowKey: prepareCheckMessage.pupilPin,
+    checkCode: prepareCheckMessage.pupil.checkCode,
+    questions: prepareCheckMessage.questions,
+    pupil: prepareCheckMessage.pupil,
+    school: prepareCheckMessage.school,
+    config: prepareCheckMessage.config,
+    tokens: {
+      sasToken: prepareCheckMessage.sasToken,
+      jwtToken: prepareCheckMessage.jwtToken
+    },
+    isCollected: false,
+    collectedAt: null,
+    hasCheckStarted: false,
+    createdAt: new Date(),
+    updatedAt: new Date()
+  }
+
+  // Happy path - write to Table Storage
+  context.log('prepareCheck: saving to table storage: ', preparedCheck)
+  context.bindings.preparedCheckTable.push(preparedCheck)
+  context.done()
+}
+
+function validateMessage (message) {
+  const check = (prop) => { if (R.not(R.has(prop, message))) { throw new Error(`Message failed validation check: missing field: ${prop}`) } }
+  const topLevelProperties = ['schoolPin', 'pupilPin', 'pupil', 'school', 'tokens', 'config', 'questions']
+  topLevelProperties.map(check)
+  validatePupil(message.pupil)
+  validateQuestions(message.questions)
+  validateSchool(message.school)
+  validateConfig(message.config)
+  return true

```

Comment:
Returns are pointless as they are never checked; it runs until an exception is thrown.

Reply:
this should be set to `context.done()`

Reply:
Actually the `return true` was written for the unit test.  Happy to reconsider.  If an error is thrown then the top level function will handle it and call `context.done(error)`